les/balance.go

```diff
@@ -0,0 +1,376 @@
+				d = ((d - time.Second) * 7 / 8) + time.Second
+			}
+			bt.nextUpdate = now + mclock.AbsTime(d)
+			bt.updateAfter(d)
+		}
+	} else {
+		bt.nextUpdate = 0
+		bt.updateAfter(0)
+	}
+}
+
+// updateAfter schedules a balance update and callback check in the future
+func (bt *balanceTracker) updateAfter(dt time.Duration) {
+	if bt.updateEvent == nil || bt.updateEvent.Cancel() {
+		if dt == 0 {
+			bt.updateEvent = nil
+		} else {
+			bt.updateEvent = bt.clock.AfterFunc(dt, func() {
+				bt.lock.Lock()
+				defer bt.lock.Unlock()
+
+				if bt.callbackCount != 0 {
+					now := bt.clock.Now()
+					bt.updateBalance(now)
+					bt.checkCallbacks(now)
+				}
+			})
+		}
+	}
+}
+
+// requestCost should be called after serving a request for the given peer
+func (bt *balanceTracker) requestCost(cost uint64) {
+	bt.lock.Lock()
+	defer bt.lock.Unlock()
+
+	if bt.stopped {
+		return
+	}
+	now := bt.clock.Now()
+	bt.updateBalance(now)
+
+	if bt.balance.pos != 0 {
+		if bt.requestFactor != 0 {
+			c := uint64(float64(cost) * bt.requestFactor)
+			if bt.balance.pos > c {
+				bt.balance.pos -= c
+			} else {
+				bt.balance.pos = 0
+			}
+			bt.checkCallbacks(now)
+		}
+	} else {
+		if bt.negRequestFactor != 0 {
+			bt.balance.neg += uint64(float64(cost) * bt.negRequestFactor)
+			bt.checkCallbacks(now)
+		}
+	}
+	bt.sumReqCost += cost
+}
+
+// getBalance returns the current positive and negative balance
+func (bt *balanceTracker) getBalance(now mclock.AbsTime) (uint64, uint64) {
+	bt.lock.Lock()
+	defer bt.lock.Unlock()
+
+	bt.updateBalance(now)
+	return bt.balance.pos, bt.balance.neg
+}
+
+// setBalance sets the positive and negative balance to the given values
+func (bt *balanceTracker) setBalance(pos, neg uint64) error {
+	bt.lock.Lock()
+	defer bt.lock.Unlock()
+
+	now := bt.clock.Now()
+	bt.updateBalance(now)
+	bt.balance.pos = pos
+	bt.balance.neg = neg
+	bt.checkCallbacks(now)
+	return nil
+}
+
+// setFactors sets the price factors. timeFactor is the price of a nanosecond of
+// connection while requestFactor is the price of a "realCost" unit.
+func (bt *balanceTracker) setFactors(neg bool, timeFactor, requestFactor float64) {
+	bt.lock.Lock()
+	defer bt.lock.Unlock()
+
+	if bt.stopped {
+		return
+	}
+	now := bt.clock.Now()
+	bt.updateBalance(now)
+	if neg {
+		bt.negTimeFactor = timeFactor
+		bt.negRequestFactor = requestFactor
+	} else {
+		bt.timeFactor = timeFactor
+		bt.requestFactor = requestFactor
+	}
+	bt.checkCallbacks(now)
+}
+
+// setCallback sets up a one-time callback to be called when priority reaches
+// the threshold. If it has already reached the threshold the callback is called
+// immediately.
+func (bt *balanceTracker) addCallback(id int, threshold int64, callback func()) {
+	bt.lock.Lock()
+	defer bt.lock.Unlock()
+
+	bt.removeCb(id)
+	idx := 0
+	for idx < bt.callbackCount && threshold < bt.callbacks[idx].threshold {
+		idx++
+	}
+	for i := bt.callbackCount; i >= idx; i-- {
```

Comment:
for i := bt.callbackCount - 1

Reply:
No, we are expanding the list and increasing callbackCount after the loop.

Reply:
fixed